quickwit/quickwit-opentelemetry/src/otlp/trace.rs

```diff
@@ -185,6 +189,99 @@ pub struct Span {
     pub links: Vec<Link>,
 }
 
+/// A wrapper around `Span` that implements `Ord` to allow insertion of spans into a `BTreeSet`.
+#[derive(Debug)]
+struct OrdSpan(Span);
+
+impl Ord for OrdSpan {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.0
+            .trace_id
+            .cmp(&other.0.trace_id)
+            .then(
+                self.0
+                    .span_start_timestamp_secs
+                    .cmp(&other.0.span_start_timestamp_secs),
+            )
+            .then(self.0.span_name.cmp(&other.0.span_name))
+            .then(self.0.span_id.cmp(&other.0.span_id))
+    }
+}
+
+impl PartialOrd for OrdSpan {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl PartialEq for OrdSpan {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.span_id == other.0.span_id
+    }
+}
+
+impl Eq for OrdSpan {}
+
+const SPAN_FINGERPRINT_SEPARATOR: char = '\0';
+
+/// Concatenation of the service name, span kind, and span name.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct SpanFingerprint(String);
+
+impl SpanFingerprint {
+    fn new(service_name: &str, span_name: &str, span_kind: i32) -> Self {
+        Self(format!(
+            "{service_name}{SPAN_FINGERPRINT_SEPARATOR}{span_name}{SPAN_FINGERPRINT_SEPARATOR}{span_kind}"
+        ))
+    }
+
+    pub fn from_string(fingerprint: String) -> Self {
+        Self(fingerprint)
+    }
+
+    pub fn span_name(&self) -> Option<&str> {
+        self.0.split(SPAN_FINGERPRINT_SEPARATOR).nth(1)
+    }
+
+    pub fn span_kind(&self) -> Option<i32> {
+        self.0
+            .split(SPAN_FINGERPRINT_SEPARATOR)
+            .nth(2)
+            .and_then(|span_kind| span_kind.parse().ok())
+    }
+
+    pub fn start_key(service_name: &str) -> Option<String> {
+        if service_name.is_empty() {
+            return None;
+        }
+        // TODO: Fix when start and end keys are base64-ish encoded.
+        let start_key = serde_json::to_string(&JsonValue::String(format!(
+            "{service_name}{SPAN_FINGERPRINT_SEPARATOR}"
+        )))
+        .expect("Start key should serialize to a JSON string without issues. ");
+        Some(start_key)
+    }
+
+    pub fn end_key(service_name: &str) -> Option<String> {
+        if service_name.is_empty() {
+            return None;
+        }
+        let mut acc = service_name.as_bytes().to_vec();
+        while let Some(byte) = acc.pop() {
+            if byte != u8::MAX {
+                acc.push(byte + 1); // Not all u32 values are valid UTF-8, but that's fine until we move to the
+                                    // bas64-based encoding.
+                let end_key = serde_json::to_string(&JsonValue::String(
+                    String::from_utf8(acc).expect("YOLO"),
+                ))
+                .expect("End key should serialize to a JSON string without issues.");
+                return Some(end_key);
+            }
+        }
+        None
```

Comment:
as `SpanFingerprint::start_key` adds an `\0` at the end of the service name, you can just push a `\u{1}` to get the end_key